tictactoe_env.py

```diff
@@ -0,0 +1,111 @@
+        turn = self.get_turn_tictactoe(state)
+        # legal_actions = np.zeros((2, 3, 3), dtype=int)
+        legal_actions = []
+        for i in range(3):
+            for j in range(3):
+                if state[:, i, j].sum() == 0:
+                    action_array = np.zeros((2, 3, 3), dtype=int)
+                    action_array[turn, i, j] = 1
+                    action_int = self.convert_action_to_int(action_array)
+                    legal_actions.append(action_int)
+        return np.array(legal_actions)
+
+
+    def is_game_over(self, state):
+        result = self.outcome(state)
+        if result == 2:
+            return 0
+        else:
+            return 1
+
+    def outcome(self, state):
+        """
+        1 is a win for x's
+        -1 is a win for o's
+        0 is a tie
+        2: the game is not over
+        """
+        for turn in range(2):
+            if all([state[turn, 0, 0] == 1, state[turn, 1, 1] == 1, state[turn, 2, 2] == 1]):
+                return self.convert_turn_to_winner(turn)
+            if all([state[turn, 2, 0] == 1, state[turn, 1, 1] == 1, state[turn, 0, 2] == 1]):
+                return self.convert_turn_to_winner(turn)
+            for i in range(3):
+                if state[turn, 0, i] == 1 and state[turn, 1, i] == 1 and state[turn, 2, i] == 1:
+                    return self.convert_turn_to_winner(turn)
+            for j in range(3):
+                if state[turn, j, 0] == 1 and state[turn, j, 1] == 1 and state[turn, j, 2] == 1:
+                    return self.convert_turn_to_winner(turn)
+        if state.sum() >= 9:
+            return 0
+        # if we reach this point, the game is not over.  return 2
+        return 2
+
+    def print_board(self, state):
+        for i in range(3):
+            row = ''
+            for j in range(3):
+                if state[0, i, j] == 1:
+                    row += 'x '
+                elif state[1, i, j] == 1:
+                    row += 'o '
+                else:
+                    row += '* '
+            print(row)
+
+    def convert_action_to_int(self, action_array):
+        """
+        action_array is 2x3x3
+        """
+        action_array = np.reshape(action_array, -1)
+        return np.where(action_array == 1)[0]
+
+    def convert_int_to_action(self, action_int):
+        action_array = np.zeros((self.action_dims), dtype=int)
+        action_array = np.reshape(action_array, -1)
+        action_array[action_int] = 1
+        action_array = np.reshape(action_array, (self.action_dims))
+        return action_array
+
+    def get_turn_tictactoe(self, state):
+        """
+        return 0 if x's turn
+        return 1 if o's turn
+        """
+        num_xs = state[0, :, :].sum()
+        num_os = state[1, :, :].sum()
+        if num_os == num_xs:
+            return 0
+        elif num_xs > num_os:
+            return 1
+        else:
+            return -1
+
+
+    def convert_turn_to_winner(self, turn):
```

Comment:
Call me silly, but how about lets make 

0 indicate x's turn, and if x won
1 indicate o's turn, and if o won.

-1 can be a draw
2 can be game is not over.

And then we can just return the `turn` directly and not do this somewhat confusing algebra

Reply:
Oh I see, this emulates the -1 and 1 of the chess environment (and what is expected by the `v` vector, I think). 

I still think this expression is overkill but whatever